url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPSystemManager.java

```diff
@@ -0,0 +1,1344 @@
+    private CopyOnWriteArrayList<Link> links;
+    /** The system state. */
+    private SystemState     state = SystemState.STOPPED;
+    /** The preferred link. */
+    private Link            preferredLink;
+    /** The system thread. */
+    private SystemThread    sysThread;
+    /** The timestamp of when the MISSING_MASTER state was entered. */
+    private long            missingMasterStart;
+    /** True if the gapping alarm is currently active. */
+    private boolean         gappingAlarmActive = false;
+    /** The watchdog to use. */
+    private GenericWatchdog watchdog;
+
+    /** The logger for this class. */
+    private static Logger log = Logger.getLogger("ocpSystemManager");
+
+    /** The alarms interface. */
+    private static AlarmInterface alarms = UtilitiesFactory
+            .getUtilInterface().getAlarmInterface();
+
+    static {
+        alarms.registerAlarm(ALARM_LINK_NOT_AVAILABLE, Link.class,
+                ALARM_LINK_NOT_AVAILABLE_ID, ALARM_LINK_NOT_AVAILABLE_SEV);
+        alarms.registerAlarm(ALARM_CANNOT_TAKE_CALLS, OCPSystemManager.class,
+                ALARM_CANNOT_TAKE_CALLS_ID, ALARM_CANNOT_TAKE_CALLS_SEV);
+    }
+
+    /**
+     * Creates a disconnected OCP System Manager using the properties specified
+     * in <tt>jOCP.properties</tt>.
+     */
+    public OCPSystemManager() {
+        FileInputStream in = null;
+        prop = new Properties();
+
+        try {
+            in = new FileInputStream("jOCP.properties");
+            prop.load(in);
+        } catch (FileNotFoundException e) {
+            log.warn("Could not find the properties file \"jOCP.properties\". "
+                    + "Falling back to defaults.", e);
+        } catch (IOException e) {
+            log.error("Error loading properties file \"jOCP.properties\". "
+                    + "Falling back to defaults.", e);
+        } finally {
+            try {
+                if (in != null) {
+                    in.close();
+                }
+            } catch (IOException e) { // CSIGNORE: EmptyBlock
+            }
+        }
+        links = new CopyOnWriteArrayList<Link>();
+        UtilitiesFactory.getUtilInterface().register(this, true);
+        sysThread = new SystemThread();
+        init();
+    }
+
+    /**
+     * Creates a disconnected OCP System Manager using the specified Properties
+     * object.
+     *
+     * @param properties
+     *            The Properties object to use
+     */
+    public OCPSystemManager(final Properties properties) {
+        this.prop = properties;
+        links = new CopyOnWriteArrayList<Link>();
+        UtilitiesFactory.getUtilInterface().register(this, true);
+        sysThread = new SystemThread();
+        init();
+    }
+
+    /**
+     * Carry out the actual initialisation work. This reads in the system-level
+     * configuration, creates all configured links, and starts the system
+     * management thread.
+     */
+    private synchronized void init() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 308-328 into a separate method.